common/shell-local/config.go

```diff
@@ -41,14 +41,17 @@ type Config struct {
 	// End dedupe with postprocessor
 	UseLinuxPathing bool `mapstructure:"use_linux_pathing"`
 
-	ctx interpolate.Context
+	// used to track the data sent to shell-local from the builder
+	// GeneratedData
+
+	ctx           interpolate.Context
+	generatedData map[string]interface{}
 }
 
 func Decode(config *Config, raws ...interface{}) error {
-	//Create passthrough for winrm password so we can fill it in once we know it
-	config.ctx.Data = &EnvVarsTemplate{
-		WinRMPassword: `{{.WinRMPassword}}`,
-	}
+	// Create passthrough for build-generated data so we can fill it in once we know
+	// it
+	config.ctx.Data = packer.BasicPlaceholderData()
```

Comment:
AH, nice !

Reply:
Does this mean the placeholder will have to know about the builder at some point ?

Reply:
I don't want the placeholders to have to know about builders. The solution I've come up with is for the builder to eventually _create_ the placeholder. I want to add a function to the builder interface, for example GetPlaceholderData(), and the that would return its own map which Packer would combine with this BasicPlaceholderData map and then pass into the Prepare() func instead of it being assigned directly the by placeholder func. 

That way, provisioners don't need to have special knowledge of what builder is running them, but still get to validate builder-specific variables so that if a provisioner is launched and is expecting to know, for example, a `.SourceAMIName` or something like that, but the builder is google compute and has created a map that doesn't include SourceAMIName, the provisioner will say "hey wait I can't do that, this value isn't going to exist even once the build has run"

I'm going to try to implement this today to give an example

Reply:
Oh nice; yeah I like this idea 🙂